src/Mod/Path/PathFeedRate.py

```diff
@@ -0,0 +1,95 @@
+# *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
+# *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
+# *   GNU Library General Public License for more details.                  *
+# *                                                                         *
+# *   You should have received a copy of the GNU Library General Public     *
+# *   License along with this program; if not, write to the Free Software   *
+# *   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  *
+# *   USA                                                                   *
+# *                                                                         *
+# ***************************************************************************
+
+import FreeCAD
+import PathScripts.PathLog as PathLog
+import PathMachineState
+import PathScripts.PathGeom as PathGeom
+import Part
+
+__title__ = "Feed Rate Helper Utility"
+__author__ = "sliptonic (Brad Collette)"
+__url__ = "https://www.freecadweb.org"
+__doc__ = "Helper for adding Feed Rate to Path Commands"
+
+"""
+TODO:  This needs to be able to handle feedrates for axes other than X,Y,Z
+"""
+
+if True:
+    PathLog.setLevel(PathLog.Level.DEBUG, PathLog.thisModule())
+    PathLog.trackModule(PathLog.thisModule())
+else:
+    PathLog.setLevel(PathLog.Level.INFO, PathLog.thisModule())
+
+
+def setFeedRate(commandlist, ToolController):
+    def _isVertical(currentposition, command):
+        x = (
+            command.Parameters["X"]
+            if "X" in command.Parameters.keys()
+            else currentposition.x
+        )
+        y = (
+            command.Parameters["Y"]
+            if "Y" in command.Parameters.keys()
+            else currentposition.y
+        )
+        z = (
+            command.Parameters["Z"]
+            if "Z" in command.Parameters.keys()
+            else currentposition.z
+        )
+        endpoint = FreeCAD.Vector(x, y, z)
+        if currentposition == endpoint:
+            return True
+        return PathGeom.isVertical(Part.makeLine(currentposition, endpoint))
+
+    feedcommands = ["G01", "G1", "G2", "G3", "G02", "G03", "G81", "G82", "G83"]
+    rapidcommands = ["G0", "G00"]
+
+    machine = PathMachineState.MachineState()
+
+    for command in commandlist:
+        if command.Name not in feedcommands + rapidcommands:
+            continue
+
+        if _isVertical(machine.getPosition(), command):
+            rate = (
+                ToolController.VertRapid.Value
+                if command in rapidcommands
+                else ToolController.VertFeed.Value
+            )
+        else:
+            rate = (
+                ToolController.HorizRapid.Value
+                if command in rapidcommands
+                else ToolController.HorizFeed.Value
+            )
+
+        params = command.Parameters
+        params["F"] = rate
+        command.Parameters = params
+
+        machine.addCommand(command)
```

Comment:
I think this would be a bit simpler and easier to understand like this. It does not use an auxiliary function nor builds too many extra objects.

```python
for command in commandlist:
    start_point = machine.getPosition()
    machine.addCommand(command)
    end_point = machine.getPosition()
    move_vec = end_point-start_point
    move_vec.z = 0 # Project onto the XY plane
    is_vertical = (move_vec.Length < PathGeom.Tolerance)
    if command.Name in feedcommands:
        if is_vertical:
            feed = ToolController.VertFeed.Value
        else:
            feed = ToolController.HorizFeed.Value
        command.Parameters["F"] = feed
    elif command.Name in rapidcommands:
        if is_vertical:
            feed = ToolController.VertRapid.Value
        else:
            feed = ToolController.HorizRapid.Value
        command.Parameters["F"] = feed
```

Also it makes sure to pass along all commands to the machine, this for example would handle correctly cases where there is a mixture of relative and absolute positioning for example (if/when the machine implements that). The vertical checking part is simple vector math, get movement vector, project to XY plane, check if there is XY motion.

A potential wrinkle is the semantics definition for the vertical feed. Some users may expect the vertical feed to be used as a limit on the Z axis speed, regardless of the angle of motion (say, a ramp). In that case we should be looking at the RATIO of motion in XY vs Z, if that ratio is is less than the ratio between horizontal and vertical feed in the TC the move should be considered vertical or not. so Z speed can never exceed the vertical feed value, regardless of how it ramps into it.

Reply:
I like it.  To clarify, the 'machine' here isn't the physical machine or even an internal representation of it.  It's a helper concept for knowing the current machine state.  Gcode is often ambiguous so given the last command, you have to read backwards, potentially to the beginning to know exactly where an axis value is.  For example given the command "G0 X10 Y5"   what is your current Z position?  Is the coolant turned on?  To know, you have to read backwards.  This object just keeps track of those kinds of things so you can query it more easily.

I don't think we should ever implement relative positioning internally either in this machinestate helper or the in the general Path Command structure.  Absolute positioning can be applied by the postprocessor if that is the output the user wants.

Reply:
Yeah, that is what I understood too. it is the gcode state machine. I still think it's a good idea to just feed the command to the machine and then ask the machine how it moved. That moves the concern of interpreting motion commands correctly to the gcode interpreter vs doing that again at this function. I could go as far as perhaps add an additional state variable to the machine class to just get the motion vector from the last command. I can really see this machinestate class being very useful in post processors.

Reply:
>  I still think it's a good idea to just feed the command to the machine and then ask the machine how it moved.

What machine?

> I could go as far as perhaps add an additional state variable to the machine class to just get the motion vector from the last command. 

A motion vector isn't enough.  That gets us our translation but not rotation elements.

Reply:
Looking at this again, there's a couple problems. 
First, with this solution, the command is added before the feedrate is applied so the machine doesn't correctly reflect the feed rate.  This can be resolved by adding the command again at the end but that seems a bit hack-y.

The command 
```
command.Parameters["F"] = feed
```
won't work.  See here:
```
>>> p = obj.Path.Commands[8]
>>> p
Command G3 [ F:0.8333 I:0 J:-2.5 K:0 X:-1.5 Y:76.038 Z:6 ]
>>> p.toGCode()
'G3 F0.833300 I0.000000 J-2.500000 K0.000000 X-1.500000 Y76.037957 Z6.000000'
>>> p.Parameters['F'] = 200
>>> p
Command G3 [ F:0.8333 I:0 J:-2.5 K:0 X:-1.5 Y:76.038 Z:6 ]
>>> p.Parameters
{'F': 200, 'I': 0.0, 'J': -2.5, 'K': 0.0, 'X': -1.5, 'Y': 76.037957, 'Z': 6.0}
>>> p.toGCode()
'G3 F0.833300 I0.000000 J-2.500000 K0.000000 X-1.500000 Y76.037957 Z6.000000'
>>> params = p.Parameters
>>> params['F'] = 200
>>> p.Parameters = params
>>> p
Command G3 [ F:200 I:0 J:-2.5 K:0 X:-1.5 Y:76.038 Z:6 ]
>>> p.toGCode()
'G3 F200.000000 I0.000000 J-2.500000 K0.000000 X-1.500000 Y76.037957 Z6.000000'
>>> 
```

Reply:
DOH, that's the problem with mutable properties in wrapped objects, seems to always be a 50/50 if they are going to work properly or not. I'm a bit tempted to call this a bug in the Path.Command object, you should be expected to modify a mutable property and have it, you know, change. The workaround using the temp variable should be ok but we should consider fixing the wrapping later. There are a few solutions I would deem more acceptable than what we got now there.

- Remove the property interface and use something like `getParameters()` and `setParameters()` instead. To make it clear that objects are passed by-copy and you cannot mutate them.
- Make `Parameters` return a proxy object so that mutating it actually does the changes expected.
- Make `Parameters` return an immutable object and have a function like `setParameter()` to be able to write to them
- Remove the extra indirection and implement the parameters dictionary on the command class itself via `__getitem__` and `__setitem__`, that way you could do `p["F"] = 200`. Also possible to implement the rest of the basic dictionary interface with `keys()` and `items()`

Reply:
Added a bug to mantis.  https://tracker.freecadweb.org/view.php?id=4831
Otherwise it's outside the scope of this PR.  I'll use the original implementation for now.  We can revisit if/when the bug is resolved.